### ISSUE TITLE
dockerTools.buildLayeredImage: layers should have a consistent order

### ISSUE THREAD

**Issue by teozkr** [NixOS/nixpkgs#53515]
## Issue description

In #47411 @grahamc states that layers no longer care about the order that they appear in. This is true for image tarballs, but for images "loaded" into the Docker daemon this depends on what driver you use.

[`overlay2`](https://docs.docker.com/storage/storagedriver/overlayfs-driver/#image-and-container-layers-on-disk) (currently the recommended driver) and ·[`device-mapper`](https://docs.docker.com/storage/storagedriver/device-mapper-driver/#image-and-container-layers-on-disk) (the fallback if `overlay2` is unavailable) both use a "snapshot" stack to implement layers, meaning that the whole stack up to that point must match in order for the cache to hit.

The only driver I could find that does *not* seem to use a snapshot stack is the deprecated [`aufs`](https://docs.docker.com/storage/storagedriver/aufs-driver/#example-image-and-container-on-disk-constructs) (but I haven't been able to verify this, since I don't have any systems with AUFS-enabled kernels).

### Steps to reproduce

1. Create a docker image with some dependencies.
```
with import <nixpkgs> {};
dockerTools.buildLayeredImage {
    name = "test-image";
    config.Cmd = writeScript "test-script"
      ''
        #!${bash}/bin/bash
        echo hi!
        # Let's add some more dependencies
        ${utillinux}/bin/lscpu
      '';
    maxLayers = 120;
}
```
2. Build and load it into the daemon (`docker load -i $(nix-build --no-out-link)`).
3. Make some trivial change to the image (such as changing the message from `hi!` to `hey!`)
4. Rebuild and reload the image

Expected: Two layers are loaded (the "top" derivation layer and the customization layer), while the rest are reused.

Actual: Almost all layers are reloaded, since the shared snapshot stack is very small.

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.20.0-arch1-1-ARCH, Arch Linux, noversion`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(teo): `"nixpkgs-19.03pre164887.4001dd679f3"`
 - nixpkgs: `/home/teo/.nix-defexpr/channels/nixpkgs`

**Comment by teozkr** [comment]
Ultimately this is a Docker/storage driver issue (the `lower` file should be tracked externally instead), but we should be able to minimize the impact by sorting the layer list according to the same popularity contest algorithm that is currently used to decide which derivations deserve to become separate layers.

**Comment by grahamc** [comment]
The layer list is sorted already in the same way. The load may not be cached but when pushing, minimal layers will be pushed. Same with fetching. At least, that has been my experience. 

Graham

**Comment by teozkr** [comment]
I haven't interacted much with registries, my problem is focused on interacting with the local daemon.

**Comment by baracoder** [comment]
Correct me if I am wrong, this is how I understand it:

So even if the order is consistent and has no effect on interaction with a registry, it might be not optimal when docker need to merge the layers.

In the worst case the most frequently changing layer is on top:
```
Layer A < - This is the project that changes 
Layer B < - Big package like JDK 
Layer C < - Other dependency
```
On `docker pull`, only A will be pulled transferred, yet while merging layers, docker will have to start from scratch and extract A, then overlay B and C on every change.

Is there some way to improve this case? Maybe an explicit option to put a certain path at the end?

**Comment by teozkr** [comment]
Exactly, @baracoder. Currently bLI uses a ["popularity contest" algorithm](https://grahamc.com/blog/nix-and-layered-docker-images) to decide which derivations should be split into separate layers when the number of derivations exceeds `maxLayers`.

According to Graham (I haven't had time to verify this) this is already used to order layers as well, but the layer order (which matters for merging) does not seem to be consistent. I suspect that this might be because of an unstable sorting algorithm somewhere.

**Comment by iwinux** [comment]
Guess: the problem may reside in `referencesByPopularity`.

With the following expression:

```nix
{
  nixpkgs ? import <nixpkgs> {}
}:

with nixpkgs;

let
  content = symlinkJoin {
    name = "xx";
    paths = [
      bashInteractive
      curl
      python37
    ];
  };

  closure = writeText "closure" "${content}";
in referencesByPopularity closure
```

If I add `python37Packages.pip` and `python37Packages.setuptools` to `paths`, we can get the diff as below:

```diff
--- before     2019-02-05 21:43:37.000000000 +0800
+++ after     2019-02-05 21:43:41.000000000 +0800
@@ -1,23 +1,25 @@
 /nix/store/fivq0nbggp4y8mhy3ixprqd7qyn1hy2j-glibc-2.27
 /nix/store/5dphwv1xs46n0qbhynny2lbhmx4xh1fc-zlib-1.2.11
-/nix/store/vs6d2fjkl4kb3jb7rwibsd76k9v2n4xy-bash-4.4-p23
 /nix/store/mi5krns5p204a3rdn1rj4ajnbqn2809i-ncurses-6.1-20181027
+/nix/store/vs6d2fjkl4kb3jb7rwibsd76k9v2n4xy-bash-4.4-p23
 /nix/store/kimqdijh6r0p8d6if8xacnijl2y4ngw5-openssl-1.0.2q
-/nix/store/s5xxfanirj8qya8aly5wwdr7b1v1bbm9-keyutils-1.6-lib
-/nix/store/hia576lm5gb8djx8pbm2lgi3q6c0kdyn-libkrb5-1.15.2
-/nix/store/ji88gwzzif3l2nplg94a615j9krh4l8m-libssh2-1.8.0
-/nix/store/447z0cjmylc3ija9b20xvvwrwkhn8nhh-nghttp2-1.35.1-lib
 /nix/store/cvjhg4cha0yh7gi2sxf31wk3h5y1s68n-bzip2-1.0.6.0.1
-/nix/store/7a19dg5dczka2yv0ai618dznvi0r6563-curl-7.63.0
 /nix/store/hqmmss40v9iqpwpmv9bgfb1kz9gaxp3z-expat-2.2.6
 /nix/store/8i618nqrx7q6zphbnn9gpkg9k3cv7iq1-gdbm-1.18.1
 /nix/store/qwmw5ddw8lk2pz4fy3irp87vj1cawvjh-libffi-3.2.1
 /nix/store/ycg4d6l6kinq7xkf687ddfvddjiwnwsr-readline-6.3p08
-/nix/store/1f4cikmcp49zs9xsl3h7km9j0var9jhx-readline-7.0p5
 /nix/store/qpa2x9g7xhmhfd12z8lrb830zidbw1b0-sqlite-3.26.0
 /nix/store/bfnknhn58j0z8sdi1980y1rbmcw61v73-xz-5.2.4
+/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2
+/nix/store/s5xxfanirj8qya8aly5wwdr7b1v1bbm9-keyutils-1.6-lib
+/nix/store/hia576lm5gb8djx8pbm2lgi3q6c0kdyn-libkrb5-1.15.2
+/nix/store/ji88gwzzif3l2nplg94a615j9krh4l8m-libssh2-1.8.0
+/nix/store/447z0cjmylc3ija9b20xvvwrwkhn8nhh-nghttp2-1.35.1-lib
+/nix/store/x54cj2603lvj6y5fmpkg7x0hvad8kp85-python3.7-setuptools-40.6.2
+/nix/store/7a19dg5dczka2yv0ai618dznvi0r6563-curl-7.63.0
+/nix/store/1f4cikmcp49zs9xsl3h7km9j0var9jhx-readline-7.0p5
 /nix/store/00aqq7zm73ca5rck3k4yf4z339lxzy2w-bash-interactive-4.4-p23
 /nix/store/k9c8xfqp467nk1b2ijagqsvczx0v1dq7-curl-7.63.0-bin
-/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2
-/nix/store/6vkjbkzfhi0znhj7bvihfcg2a3j324ny-xx
-/nix/store/bb84sazs3429sjnbvq203586k14sg71h-closure
+/nix/store/qmd9d1f419lxkg0mr2wdg7cv1y8h31fm-python3.7-pip-18.1
+/nix/store/2psfxz41m4fc7zw3sj1yzm7424r5h463-xx
+/nix/store/p374d8a6hb19i3a9hv0aymm4mba8h5nw-closure
```

At the very least, `/nix/store/vs6d2fjkl4kb3jb7rwibsd76k9v2n4xy-bash-4.4-p23` should not be reordered after I add these two paths.

**Comment by grahamc** [comment]
They are reordered because the popularity weights change.

**Comment by memberbetty** [comment]
I am not sure how this relates to the amount of data being uploaded to a registry, but when I make a one character change, the upload process still uploads the size of the image. (checked via bandwidth monitoring tools).

My expectation is that only a single layer would be uploaded.

**Comment by teozkr** [comment]
@memberbetty: What tool (and version) do you use to upload the images? IIRC, `docker push` will check for dupe layers before loading, but `docker load` won't. Skopeo will check both Daemon and Registry targets, but only in very recent versions (0.1.33 and later, I think).

**Comment by memberbetty** [comment]
@teozkr `docker push`

Docker version 18.09.2, build 62479626f213818ba5b4565105a05277308587d5

**Comment by danieldk** [comment]
I am running into the same problem, some containers share a 3GB data file. In the generated tarball, they have the same hashes. But there is no layer sharing after a `docker load`, nor after `docker push`ing the resulting images to Docker Hub. Tested with the `device-mapper` and `overlay2` storage drivers.

**Comment by danieldk** [comment]
Found the culprit and submitted a PR: #62719.

**Comment by nlewo** [comment]
Fixed by the merged PR #62719.
@danieldk tip: you can add `Fixes #ISSUE` in your commit message to automatically close issues (see https://github.blog/2013-01-22-closing-issues-via-commit-messages/).

Thanks.

**Comment by teozkr** [comment]
@nlewo That's a different issue (although it is a prerequisite for this issue's relevance). #62719 ensures that layer tarballs are identical so that the layers are deduplicated at the Docker Registry. However, this issue is about that the Docker _Daemon_ will only deduplicate layers if the whole layer stack up to that point matches, since it doesn't differentiate between the layer diff and the merged overlay.
